lock-api/src/main/java/com/palantir/lock/watch/LockWatchReferences.java

```diff
@@ -0,0 +1,198 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.watch;
+
+import org.immutables.value.Value;
+
+import com.fasterxml.jackson.annotation.JsonSubTypes;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import com.fasterxml.jackson.annotation.JsonTypeName;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import com.google.common.collect.Range;
+import com.palantir.lock.AtlasLockDescriptorRanges;
+import com.palantir.lock.LockDescriptor;
+
+public final class LockWatchReferences {
+    private LockWatchReferences() {
+        // no
+    }
+
+    public static LockWatchReference entireTable(String qualifiedTableRef) {
+        return ImmutableEntireTable.of(qualifiedTableRef);
+    }
+
+    public static LockWatchReference rowPrefix(String qualifiedTableRef, byte[] rowPrefix) {
+        return ImmutableRowPrefix.of(qualifiedTableRef, rowPrefix);
+    }
+
+    public static LockWatchReference rowRange(String qualifiedTableRef, byte[] startInclusive, byte[] endExclusive) {
+        return ImmutableRowRange.of(qualifiedTableRef, startInclusive, endExclusive);
+    }
+
+    public static LockWatchReference exactRow(String qualifiedTableRef, byte[] row) {
+        return ImmutableExactRow.of(qualifiedTableRef, row);
+    }
+
+    public static LockWatchReference exactCell(String qualifiedTableRef, byte[] row, byte[] col) {
+        return ImmutableExactCell.of(qualifiedTableRef, row, col);
+    }
+
+    @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
+    @JsonSubTypes({
+            @JsonSubTypes.Type(value = ImmutableEntireTable.class, name = EntireTable.TYPE),
+            @JsonSubTypes.Type(value = ImmutableRowPrefix.class, name = RowPrefix.TYPE),
+            @JsonSubTypes.Type(value = ImmutableRowRange.class, name = RowRange.TYPE),
+            @JsonSubTypes.Type(value = ImmutableExactRow.class, name = ExactRow.TYPE),
+            @JsonSubTypes.Type(value = ImmutableExactCell.class, name = ExactCell.TYPE)})
+    public interface LockWatchReference {
+        <T> T accept(Visitor<T> visitor);
+    }
+
+    public interface Visitor<T> {
```

Comment:
Discussed offline. This is not exactly what I would expect in terms of the class hierarchy, but is fine and we can change later on if needed.